src/mesh/rbf.f90

```diff
@@ -0,0 +1,597 @@
+INTEGER,ALLOCATABLE    :: HexaMapLinear(:,:)
+REAL                   :: xTriLinear(3,BoundaryOrder**3),xBiLinear(3,BoundaryOrder**2)
+REAL                   :: xCornerVol(1:3,8),xCornerSurf(1:3,4)
+REAL                   :: dist,x(3),xTmp(3),rbfvalue
+INTEGER                :: iBP
+REAL                   :: xMin,xMax,yMin,yMax,xBary(3)
+!===================================================================================================================================
+
+WRITE(UNIT_StdOut,'(132("-"))')
+WRITE(UNIT_stdOut,'(A)') ' RBF VOLUME CURVING'
+
+! Box where curving should be done
+xMin = xlim(1,iRBFBox)
+xMax = xlim(2,iRBFBox)
+yMin = ylim(1,iRBFBox)
+yMax = ylim(2,iRBFBox)
+
+! Build 2D Vandermonde from corner nodes to bi-linear side nodes
+CALL getQuadBasis(1,BoundaryOrder,Vdm_SurfBiLinear,D_tmp)
+DEALLOCATE(D_tmp)
+
+! Build 3D Vandermonde from corner nodes to tri-linear side nodes
+CALL getHexaBasis(1,BoundaryOrder,Vdm_VolTriLinear,D_tmp)
+DEALLOCATE(D_tmp)
+
+! Build mapping between 1D and tensor product ordering for a bi-linear surface
+CALL getBasisMappingQuad(1,nAns,QuadMapLinear,QuadMapInvLinear)
+DEALLOCATE(QuadMapLinear)
+
+! Build mapping between 1D and tensor product ordering for a bi-linear surface
+CALL getBasisMappingHexa(1,nAns,HexaMapLinear,HexaMapInvLinear)
+DEALLOCATE(HexaMapLinear)
+
+
+!--------------------------------------------------------------------------------------------------------!
+! Count number of control points, needed to allocate space etc.
+!--------------------------------------------------------------------------------------------------------!
+
+! Reset the tmp variable later used to mark already visited control points
+Elem=>FirstElem
+DO WHILE(ASSOCIATED(Elem))
+  Side=>Elem%firstSide
+  DO WHILE(ASSOCIATED(Side))
+    IF(.NOT.ASSOCIATED(Side%BC)) THEN
+      ! Go through all 4 corner nodes
+      Side%CurvedNode(QuadMapInv(0,0))%np%tmp = 0
+      Side%CurvedNode(QuadMapInv(0,N))%np%tmp = 0
+      Side%CurvedNode(QuadMapInv(N,0))%np%tmp = 0
+      Side%CurvedNode(QuadMapInv(N,N))%np%tmp = 0
+    ELSE IF(Side%BC%BCType.EQ.1) THEN
+      ! Go through all 4 corner nodes
+      Side%CurvedNode(QuadMapInv(0,0))%np%tmp = 0
+      Side%CurvedNode(QuadMapInv(0,N))%np%tmp = 0
+      Side%CurvedNode(QuadMapInv(N,0))%np%tmp = 0
+      Side%CurvedNode(QuadMapInv(N,N))%np%tmp = 0
+    ELSE
+      DO i=1,Side%nCurvedNodes
+        Side%CurvedNode(i)%np%tmp = 0
+      END DO
+    END IF
+    Side=>Side%nextElemSide
+  END DO
+  Elem=>Elem%nextElem
+END DO
+
+nBP = 0
+Elem=>FirstElem
+DO WHILE(ASSOCIATED(Elem))
+  Side=>Elem%firstSide
+  DO WHILE(ASSOCIATED(Side))
+    IF(.NOT.ASSOCIATED(Side%BC)) THEN
+      ! Not a BC, include the corner nodes
+      IF (.NOT.(Side%CurvedNode(QuadMapInv(0,0))%np%tmp.EQ.1)) THEN
+        ! Check if node is inside the box
+        x = Side%CurvedNode(QuadMapInv(0,0))%np%x
+        IF ((x(1).LT.xMin).OR.(x(1).GT.xMax).OR.(x(2).LT.yMin).OR.(x(2).GT.yMax)) EXIT
```

Comment:
Do we want to perform this check on a node or an element basis? Given the choice, I'd prefer the latter since I see no guarantees on the amount of displacement? We might produce invalid elements this way.

Reply:
There is a branch which introduces an additional inner box, which also changes this section. I'm not sure what the intention behind it was

Reply:
After discussion, we leave it as is. Could yield invalid elements, but most likely only if the displacements is way too large for the given number of elements, since the displacements are guaranteed to be smoothly interpolated to the nodes by the RBFs.